### ISSUE TITLE
gio/System/GIO/File/IOError.chs does not compile under recent MSYS2

### ISSUE THREAD

**Issue by Ralith** [gtk2hs/gtk2hs#153]
Configuring gio-0.13.1.1...
    Preprocessing library gio-0.13.1.1...

    C:\msys64\home\Ralith\votecount\gtk2hs\gio\System\GIO\File\IOError.chs:49:15: parse error on input `*'

This can be worked around by deleting some comments:

```diff
diff --git a/gio/System/GIO/File/IOError.chs b/gio/System/GIO/File/IOError.chs
index 6890085..f53ed2e 100644
--- a/gio/System/GIO/File/IOError.chs
+++ b/gio/System/GIO/File/IOError.chs
@@ -46,8 +46,7 @@ import System.Glib.FFI
 --------------------
 -- Methods

--- | Converts errno.h error codes into GIO error codes.
-ioErrorFromErrno :: Int  -- ^ @err@ Error number as defined in errno.h.
-                 -> IO IOErrorEnum  -- ^ returns 'IOErrorEnum' value for the given errno.h error number.
+ioErrorFromErrno :: Int
+                 -> IO IOErrorEnum
 ioErrorFromErrno err =
   liftM (toEnum . fromIntegral) $ {#call g_io_error_from_errno#} (fromIntegral err)
```
Presumably this could be better fixed by tweaking the comments somehow, but it's not clear to me what exactly is causing the issue.

This is using the mingw64 environment up to date as of yesterday, having gcc 5.3.0.

**Comment by RyanGlScott** [comment]
This is a case of CPP being too aggressive. On most OSes, `errno.h` simply defined `errno` as `external int errno`, but in MinGW-w64, it's defined as a macro (in `mingw/x86_64-w64-mingw32/include/errno.h`):

```c
_CRTIMP extern int *__cdecl _errno(void);
#define errno (*_errno())
```

Unfortunately, this means that `c2hs` or `gtk2hs-buildtools` will create this code for `System/GIO/File/IOError.hs` erroneously:

```haskell
-- | Converts @
{-# LINE 49 ".\\\\System\\\\GIO\\\\File\\\\IOError.chs" #-}
              (*_errno())
{-# LINE 49 ".\\\\System\\\\GIO\\\\File\\\\IOError.chs" #-}
                   .h@ error codes into GIO error codes.
ioErrorFromErrno :: Int -- ^ @err@ Error number as defined in
{-# LINE 50 ".\\\\System\\\\GIO\\\\File\\\\IOError.chs" #-}
                                                              (*_errno())
{-# LINE 50 ".\\\\System\\\\GIO\\\\File\\\\IOError.chs" #-}
                                                                   .h.
                 -> IO IOErrorEnum -- ^ returns 'IOErrorEnum' value for the given
{-# LINE 51 ".\\\\System\\\\GIO\\\\File\\\\IOError.chs" #-}
                                                                                  (*_errno())
{-# LINE 51 ".\\\\System\\\\GIO\\\\File\\\\IOError.chs" #-}
                                                                                       .h error number.
ioErrorFromErrno err =
  liftM (toEnum . fromIntegral) $ g_io_error_from_errno (fromIntegral err)
```

I'm not sure what the cleanest way to fix it is. One possible solution is to `#undef errno` like so:

```diff
diff --git a/gio/System/GIO/File/IOError.chs b/gio/System/GIO/File/IOError.chs
index 6890085..1fbfcb2 100644
--- a/gio/System/GIO/File/IOError.chs
+++ b/gio/System/GIO/File/IOError.chs
@@ -46,6 +46,7 @@ import System.Glib.FFI
 --------------------
 -- Methods

+#undef errno
 -- | Converts errno.h error codes into GIO error codes.
 ioErrorFromErrno :: Int  -- ^ @err@ Error number as defined in errno.h.
                  -> IO IOErrorEnum  -- ^ returns 'IOErrorEnum' value for the given errno.h error number.

```

**Comment by RyanGlScott** [comment]
Or better yet, there's an even easier fix:

```diff
diff --git a/gio/System/GIO/File/IOError.chs b/gio/System/GIO/File/IOError.chs
index 6890085..51ee03c 100644
--- a/gio/System/GIO/File/IOError.chs
+++ b/gio/System/GIO/File/IOError.chs
@@ -1,4 +1,3 @@
-{-# LANGUAGE CPP #-}
 --  GIMP Toolkit (GTK) Binding for Haskell: binding to gio -*-haskell-*-
 --
 --  Author : Andy Stewart

```

This module doesn't require `CPP` at all (just `gtk2hs-buildtools`), so we can avoid this issue altogether by removing it. I've opened a #156 to address this.

**Comment by RyanGlScott** [comment]
This can be closed, as it was fixed by #156.
